File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

	static List<int[]> chicken;
	static List<int[]> house;
	static int N, R;
	static int[] nums;
	static int answer;

	public static void main(String[] args) throws IOException {
		StringBuilder sb = new StringBuilder();

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		R = Integer.parseInt(st.nextToken());
		nums = new int[R];

		chicken = new ArrayList<int[]>();
		house = new ArrayList<int[]>();

		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < N; j++) {
				int n = Integer.parseInt(st.nextToken());
				if (n == 1)
					house.add(new int[] { i, j });
				if (n == 2)
					chicken.add(new int[] { i, j });
			}
		}
		answer = 2 * N * house.size();

		// 입력완 로직시작

		combi(0, 0);
		System.out.println(answer);

	}

	private static void combi(int depth, int start) {
		if (depth == R) {
			int d = distAll();
			answer = d < answer ? d : answer;
			return;
		}
		for (int i = start; i < chicken.size(); i++) {
			nums[depth] = i;
			combi(depth + 1, i + 1);
			nums[depth] = 0;
		}

	}

	private static int distAll() {
		int result = 0;
		for (int[] h : house) {
			result += dist(h);
		}
		return result;
	}

	private static int dist(int[] n1) {
		int result = 2 * N;
		for (int n : nums) {
			int d = Math.abs(n1[0] - chicken.get(n)[0]) + Math.abs(n1[1] - chicken.get(n)[1]);
			result = d < result ? d : result;
		}
		return result;
	}

}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
# 시작시간 : 14시 43분
# 제출시간 : 15시 4분
# 제출횟수 : 1회


# 입력받기 #################################
N, M = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(N)]

# 치킨 및 집의 인덱스 뽑기 #######################
chicken = []
home = []

for r in range(N):
    for c in range(N):
        if arr[r][c] == 1:
            home.append((r, c))
        elif arr[r][c] == 2:
            chicken.append((r, c))

C, H = len(chicken), len(home)


# 치킨집과 집 사이 거리행렬 만들기 ##########################
def clac_dist(ci, hi):
    return abs(chicken[ci][0] - home[hi][1])


dist = [[0] * H for i in range(C)]  # 행 : 치킨, 열: 집
for r in range(C):
    for c in range(H):
        dist[r][c] = abs(chicken[r][0] - home[c][0]) + abs(chicken[r][1] - home[c][1])  # ci번쨰 치킨집과 hi번쨰 집의 거리


# 치킨집 C개 중 M개 뽑기
def combi(depth, start):
    if depth == M:  # 종료조건
        ans[0] = min(ans[0], calc_city_dist())  # 정답 갱신
        return
    for i in range(start, C):
        nums[depth] = i
        combi(depth + 1, i + 1)


# M개를 선택했을 떄 도시의 치킨거리
def calc_city_dist():
    result = 0  # 도시의 치킨 거리
    for i in range(H):
        mn = 1000000  # i번쨰 집의 치킨 거리(최대 100)
        for n in nums:
            mn = min(mn, dist[n][i])
        result += mn
    return result


ans = [10000000000]  # 예상 최대 100 * 100
nums = [0] * M
combi(0, 0)
print(ans[0])